akka-http-core/src/main/scala/akka/http/scaladsl/settings/ConnectionPoolSettings.scala

```diff
@@ -52,38 +85,81 @@ abstract class ConnectionPoolSettings extends js.ConnectionPoolSettings { self:
 
   // ---
 
-  // overrides for more precise return type
-  override def withMaxConnections(n: Int): ConnectionPoolSettings = self.copy(maxConnections = n)
-  override def withMinConnections(n: Int): ConnectionPoolSettings = self.copy(minConnections = n)
-  override def withMaxRetries(n: Int): ConnectionPoolSettings = self.copy(maxRetries = n)
-  override def withMaxOpenRequests(newValue: Int): ConnectionPoolSettings = self.copy(maxOpenRequests = newValue)
-  override def withPipeliningLimit(newValue: Int): ConnectionPoolSettings = self.copy(pipeliningLimit = newValue)
-  override def withBaseConnectionBackoff(newValue: FiniteDuration): ConnectionPoolSettings = self.copy(baseConnectionBackoff = newValue)
-  override def withMaxConnectionBackoff(newValue: FiniteDuration): ConnectionPoolSettings = self.copy(maxConnectionBackoff = newValue)
-
-  override def withIdleTimeout(newValue: Duration): ConnectionPoolSettings = self.copy(idleTimeout = newValue)
-  override def withMaxConnectionLifetime(newValue: Duration): ConnectionPoolSettings = self.copy(maxConnectionLifetime = newValue)
+  @ApiMayChange
+  def withHostOverrides(hostOverrides: immutable.Seq[(String, ConnectionPoolSettings)]): ConnectionPoolSettings = self.copy(hostOverrides = hostOverrides.map { case (h, s) => ConnectionPoolSettingsImpl.hostRegex(h) -> s })
 
-  // overloads for idiomatic Scala use
-  def withConnectionSettings(newValue: ClientConnectionSettings): ConnectionPoolSettings = self.copy(connectionSettings = newValue)
+  @ApiMayChange
+  def appendHostOverride(hostPattern: String, settings: ConnectionPoolSettings): ConnectionPoolSettings = self.copy(hostOverrides = hostOverrides :+ (ConnectionPoolSettingsImpl.hostRegex(hostPattern) -> settings))
+
+  override def withMaxConnections(n: Int): ConnectionPoolSettings = self.copy(maxConnections = n, hostOverrides = hostOverrides.map { case (k, v) => k -> v.withMaxConnections(n) })
```

Comment:
I think we should create an alternative copy method that provides part of that functionality to keep this code DRYer. E.g.

```scala
def copyAndAdaptOverrides(/* same parameters as copy here but manually provided */, updateOverrides: ConnectionPoolSettings => ConnectionPoolSettings)
```

Then this would be

```scala
self.copyAndAdaptOverrides(maxConnections = n, _.withMaxConnections(n))
```

Reply:
agreed, adding a (private) `copyDeep` makes this a bit nicer